app/ante/reject_msgs.go

```diff
@@ -1,48 +1,143 @@
+func NewRejectMessagesDecorator(cdc codec.Codec, groupKeeper *groupkeeper.Keeper, disabledMsgTypeURLs ...string) RejectMessagesDecorator {
+	disabledMsgsMap := make(map[string]struct{})
+	for _, url := range disabledMsgTypeURLs {
+		disabledMsgsMap[url] = struct{}{}
+	}
+
 	return RejectMessagesDecorator{
-		disabledMsgTypeURLs: []string{},
+		disabledMsgTypeURLs: disabledMsgsMap,
+		cdc:                 cdc,
+		groupKeeper:         groupKeeper,
 	}
 }
 
-// AnteHandle rejects messages that requires ethereum-specific authentication.
-// For example `MsgEthereumTx` requires fee to be deducted in the antehandler in
+// AnteHandle recursively rejects messages such as those that requires ethereum-specific authentication.
+// For example `MsgEthereumTx` requires fee to be deducted in the ante handler in
 // order to perform the refund.
-func (rmd RejectMessagesDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
-	for _, msg := range tx.GetMsgs() {
+func (rmd RejectMessagesDecorator) AnteHandle(
+	ctx sdk.Context,
+	tx sdk.Tx,
+	simulate bool,
+	next sdk.AnteHandler,
+) (sdk.Context, error) {
+	if err := rmd.checkMsgs(ctx, tx.GetMsgs(), 0); err != nil {
+		return ctx, errorsmod.Wrapf(sdkerrors.ErrUnauthorized, err.Error())
+	}
+	return next(ctx, tx, simulate)
+}
+
+const maxNestedMsgs = 6
+
+func (rmd RejectMessagesDecorator) checkMsgs(ctx sdk.Context, msgs []sdk.Msg, nestedMsgs int) error {
+	if nestedMsgs >= maxNestedMsgs {
+		return fmt.Errorf("found more nested msgs than permitted. Limit is : %d", maxNestedMsgs)
+	}
+
+	for _, msg := range msgs {
+		typeURL := sdk.MsgTypeURL(msg)
+		if rmd.isDisabledMsg(typeURL) {
+			return fmt.Errorf("found disabled msg type: %s", typeURL)
+		}
+
 		if _, ok := msg.(*evmtypes.MsgEthereumTx); ok {
-			return ctx, errorsmod.Wrapf(
+			return errorsmod.Wrapf(
 				sdkerrors.ErrInvalidType,
 				"MsgEthereumTx needs to be contained within a tx with 'ExtensionOptionsEthereumTx' option",
 			)
 		}
 
-		typeURL := sdk.MsgTypeURL(msg)
-		for _, disabledTypeURL := range rmd.disabledMsgTypeURLs {
-			if typeURL == disabledTypeURL {
-				return ctx, errorsmod.Wrapf(
-					sdkerrors.ErrUnauthorized,
-					"MsgTypeURL %s not supported",
-					typeURL,
-				)
+		switch concreteMsg := msg.(type) {
+		case *authz.MsgExec:
+			innerMsgs, err := concreteMsg.GetMessages()
+			if err != nil {
+				return err
+			}
+			nestedMsgs++
+			if err := rmd.checkMsgs(ctx, innerMsgs, nestedMsgs); err != nil {
+				return err
+			}
+		case *authz.MsgGrant:
+			authorization, err := concreteMsg.GetAuthorization()
+			if err != nil {
+				return err
 			}
+			nestedMsgs++
+			url := authorization.MsgTypeURL()
+			if rmd.isDisabledMsg(url) {
+				return fmt.Errorf("granting disabled msg type: %s is not allowed", url)
+			}
+		case *govtypesv1.MsgSubmitProposal:
+			nestedMsgs++
+			if err := rmd.checkMessages(ctx, concreteMsg.Messages, nestedMsgs); err != nil {
```

Comment:
why not using `concreteMsg.GetMsgs()` as in `authz.MsgExec`, to reuse same interface?